tests/error/poly_errors/non_linear2.err

```diff
@@ -3,5 +3,5 @@ Guppy compilation failed. Error in file $FILE:23
 21:    @guppy(module)
 22:    def main() -> None:
 23:        foo(h)
-           ^^^^^^
-GuppyTypeError: Cannot instantiate non-linear type variable `T` in type `forall T. (T -> T) -> None` with linear type `qubit`
+               ^
+GuppyTypeError: Expected argument of type `?T -> ?T`, got `qubit -> qubit`
```

Comment:
Why did this error get worse?

Reply:
I think it's failing comparing `T` with `NoFlags` against `qubit` with `Inout` before it can get to the linearity check. I'll add a better error message for these flag mismatches

Reply:
BUT the case I was thinking of in my first comment _also_ happens. I've just added `poly_errors/non_linear3.py` to demonstrate this. `@inout` flags are silent during printing because they're not user-facing, but they still mess up unification. I'll add an issue for this

Reply:
Added #487